src/Microsoft.ML.Transforms/CustomMappingTransformerFactory.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.Data.DataView;
+
+namespace Microsoft.ML.Transforms
+{
+    /// <summary>
+    /// Place this attribute onto a type to cause it to be considered a custom mapping transformer factory.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
+    public sealed class CustomMappingTransformerFactoryAttribute : ExtensionBaseAttribute
+    {
+        public CustomMappingTransformerFactoryAttribute(string contractName)
+            : base(contractName)
+        {
+        }
+    }
+
+    internal interface ICustomMappingTransformerFactory
+    {
+        ITransformer CreateTransformerObject(IHostEnvironment env, string contractName);
+    }
+
+    /// <summary>
+    /// The base type for custom mapping transformer factories.
+    /// </summary>
+    /// <typeparam name="TSrc">The type that describes what 'source' columns are consumed from the input <see cref="IDataView"/>.</typeparam>
+    /// <typeparam name="TDst">The type that describes what new columns are added by this transform.</typeparam>
+    public abstract class CustomMappingTransformerFactory<TSrc, TDst> : ICustomMappingTransformerFactory
+        where TSrc : class, new()
+        where TDst : class, new()
+    {
+        public abstract Action<TSrc, TDst> GetTransformer();
```

Comment:
>GetTransformer [](start = 43, length = 14)

You may be disinclined to do it now, and maybe someone else should do it, but this is definitely something we'll want to document very carefully. It is a bit odd, this pattern of mutable input as a return value. (I of course know exactly why we do it, but a user will not.) Not saying you should fix it here, just registering something we'll "eventually" need to fix.

Reply:
Agreed.  Logged #2657.